File: superagentweb-core/src/main/java/com/siyee/superagentweb/utils/SuperAgentWebUtils.java
package com.siyee.superagentweb.utils;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.text.TextUtils;
import android.view.ViewGroup;
import android.webkit.WebView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.core.app.AppOpsManagerCompat;
import androidx.core.content.ContextCompat;

import java.io.File;
import java.math.BigInteger;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @author hackycy
 * @date 2020/8/3
 */
public class SuperAgentWebUtils {

    private static Toast mToast = null;

    private static Handler mHandler = null;

    private SuperAgentWebUtils() {
        throw new UnsupportedOperationException("SuperAgentWebUtils can' t init");
    }

    public static void clearWebView(WebView m) {
        if (m == null) {
            return;
        }
        if (Looper.myLooper() != Looper.getMainLooper()) {
            return;
        }
        m.loadUrl("about:blank");
        m.stopLoading();
        if (m.getHandler() != null) {
            m.getHandler().removeCallbacksAndMessages(null);
        }
        m.removeAllViews();
        ViewGroup mViewGroup = null;
        if ((mViewGroup = ((ViewGroup) m.getParent())) != null) {
            mViewGroup.removeView(m);
        }
        m.setWebChromeClient(null);
        m.setWebViewClient(null);
        m.setTag(null);
        m.clearHistory();
        m.destroy();
        m = null;
    }

    public static String getMIMEType(File f) {
        String type = "";
        String fName = f.getName();
        /* 取得扩展名 */
        String end = fName.substring(fName.lastIndexOf(".") + 1, fName.length()).toLowerCase();
        /* 依扩展名的类型决定MimeType */
        if (end.equals("pdf")) {
            type = "application/pdf";//
        } else if (end.equals("m4a") || end.equals("mp3") || end.equals("mid") ||
                end.equals("xmf") || end.equals("ogg") || end.equals("wav")) {
            type = "audio/*";
        } else if (end.equals("3gp") || end.equals("mp4")) {
            type = "video/*";
        } else if (end.equals("jpg") || end.equals("gif") || end.equals("png") ||
                end.equals("jpeg") || end.equals("bmp")) {
            type = "image/*";
        } else if (end.equals("apk")) {
            type = "application/vnd.android.package-archive";
        } else if (end.equals("pptx") || end.equals("ppt")) {
            type = "application/vnd.ms-powerpoint";
        } else if (end.equals("docx") || end.equals("doc")) {
            type = "application/vnd.ms-word";
        } else if (end.equals("xlsx") || end.equals("xls")) {
            type = "application/vnd.ms-excel";
        } else {
            type = "*/*";
        }
        return type;
    }

    public static boolean hasPermission(@NonNull Context context, @NonNull String... permissions) {
        return hasPermission(context, Arrays.asList(permissions));
    }

    public static boolean hasPermission(@NonNull Context context, @NonNull List<String> permissions) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            return true;
        }
        for (String permission : permissions) {
            int result = ContextCompat.checkSelfPermission(context, permission);
            if (result == PackageManager.PERMISSION_DENIED) {
                return false;
            }
            String op = AppOpsManagerCompat.permissionToOp(permission);
            if (TextUtils.isEmpty(op)) {
                continue;
            }
            result = AppOpsManagerCompat.noteProxyOp(context, op, context.getPackageName());
            if (result != AppOpsManagerCompat.MODE_ALLOWED) {
                return false;
            }
        }
        return true;
    }

    public static List<String> getDeniedPermissions(Activity activity, String[] permissions) {
        if (permissions == null || permissions.length == 0) {
            return null;
        }
        List<String> deniedPermissions = new ArrayList<>();
        for (int i = 0; i < permissions.length; i++) {
            if (!hasPermission(activity, permissions[i])) {
                deniedPermissions.add(permissions[i]);
            }
        }
        return deniedPermissions;
    }

    public static boolean isUIThread() {
        return Looper.myLooper() == Looper.getMainLooper();
    }

    public static void runInUiThread(Runnable runnable) {
        if (mHandler == null) {
            mHandler = new Handler(Looper.getMainLooper());
        }
        mHandler.post(runnable);
    }

    @SuppressLint("ShowToast")
    static void toastShowShort(Context context, String msg) {
        if (mToast == null) {
            mToast = Toast.makeText(context.getApplicationContext(), msg, Toast.LENGTH_SHORT);
        } else {
            mToast.setText(msg);
        }
        mToast.show();
    }

    public static String md5(String str) {
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            md.update(str.getBytes());
            return new BigInteger(1, md.digest()).toString(16);
        } catch (Exception e) {
            if (LogUtils.isDebug()) {
                e.printStackTrace();
            }
        }
        return "";
    }

    public static String getApplicationName(Context context) {
        PackageManager packageManager = null;
        ApplicationInfo applicationInfo = null;
        String applicationName = "";
        try {
            packageManager = context.getApplicationContext().getPackageManager();
            applicationInfo = packageManager.getApplicationInfo(context.getPackageName(), 0);
            applicationName = (String) packageManager.getApplicationLabel(applicationInfo);
        } catch (Exception ignored) {}
        return applicationName;
    }

    public static boolean checkWifi(Context context) {
        ConnectivityManager connectivity = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        if (connectivity == null) {
            return false;
        }
        @SuppressLint("MissingPermission") NetworkInfo info = connectivity.getActiveNetworkInfo();
        return info != null && info.isConnected() && info.getType() == ConnectivityManager.TYPE_WIFI;
    }

    public static boolean checkNetwork(Context context) {
        ConnectivityManager connectivity = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        if (connectivity == null) {
            return false;
        }
        @SuppressLint("MissingPermission") NetworkInfo info = connectivity.getActiveNetworkInfo();
        return info != null && info.isConnected();
    }

}

File: superagentweb-core/src/main/java/com/siyee/superagentweb/SuperAgentWebUtils.kt
package com.siyee.superagentweb

import android.content.Context
import android.os.Environment
import android.os.Looper
import android.text.TextUtils
import android.view.ViewGroup
import android.webkit.WebView
import java.io.File

fun cleanWebView(m: WebView?) {
    if (m == null) {
        return
    }
    if (Looper.myLooper() != Looper.getMainLooper()) {
        return
    }
    m.loadUrl("about:blank")
    m.stopLoading()
    if (m.handler != null) {
        m.handler.removeCallbacksAndMessages(null)
    }
    m.removeAllViews()
    val mViewGroup: ViewGroup? = m.parent as? ViewGroup
    mViewGroup?.removeView(m)
    m.setWebChromeClient(null)
    m.webViewClient = null
    m.tag = null
    m.clearHistory()
    m.destroy()
}

fun getSuperAgentWebFilePath(context: Context): String {
    if (!TextUtils.isEmpty(SUPERAGENTWEB_FILE_PATH)) {
        return SUPERAGENTWEB_FILE_PATH
    }
    return ""
}

fun getDiskExternalCacheDir(context: Context): String? {
    val file = context.externalCacheDir
    if (Environment.MEDIA_MOUNTED.equals(Environment.getStorageState(file))) {
        return file?.absolutePath
    }
    return null
}

fun getMIMEType(file: File): String {
    var type = ""
    val fName: String = file.getName()
    /* 取得扩展名 */
    /* 取得扩展名 */
    val end = fName.substring(fName.lastIndexOf(".") + 1, fName.length).toLowerCase()
    /* 依扩展名的类型决定MimeType */
    /* 依扩展名的类型决定MimeType */type = if (end == "pdf") {
        "application/pdf" //
    } else if (end == "m4a" || end == "mp3" || end == "mid" || end == "xmf" || end == "ogg" || end == "wav") {
        "audio/*"
    } else if (end == "3gp" || end == "mp4") {
        "video/*"
    } else if (end == "jpg" || end == "gif" || end == "png" || end == "jpeg" || end == "bmp") {
        "image/*"
    } else if (end == "apk") {
        "application/vnd.android.package-archive"
    } else if (end == "pptx" || end == "ppt") {
        "application/vnd.ms-powerpoint"
    } else if (end == "docx" || end == "doc") {
        "application/vnd.ms-word"
    } else if (end == "xlsx" || end == "xls") {
        "application/vnd.ms-excel"
    } else {
        "*/*"
    }
    return type
}

fun isUIThread(): Boolean {
    return Looper.myLooper() == Looper.getMainLooper()
}